### PR TITLE
Specify HTTP(S) DNA location in conductor config

### PR DESCRIPTION
**Pull request by maackle**
## PR summary

- Changes DnaConfiguration::file to DnaConfiguration::location, deprecating the former but aliasing it to the latter
- Adds DnaLocation type which represents either an HTTP URL or a file location, uses this everywhere a DNA location is referenced
- Performs download from URL if specified, when attempting to get Content from the DnaConfig

## testing/benchmarking notes

No tests written for the new download functionality

## changelog

- [x] if this is a code change that effects some consumer (e.g. zome developers) of holochain core,  then it has been added to [our between-release changelog](https://github.com/holochain/holochain-rust/blob/develop/CHANGELOG-UNRELEASED.md) with the format 

```markdown
- summary of change [PR#1234](https://github.com/holochain/holochain-rust/pull/1234)
```

## documentation

- [ ] this code has been documented according to our [docs checklist](https://hackmd.io/@freesig/Hk9AmKJNS)

### CODE REVIEW

File: crates/conductor_lib/src/conductor/admin.rs [hunk 1]
```diff
@@ -903,13 +905,15 @@ id = 'new-dna'"#,
             vec![
                 DnaConfiguration {
                     id: String::from("test-dna"),
-                    file: String::from("app_spec.dna.json"),
+                    file: None,
```
Comment (zippy) [inline crates/conductor_lib/src/conductor/admin.rs]: lets add a comment onto these `file: None` about the deprecation so it's clear in the code that they are going away.

File: crates/conductor_lib/src/conductor/base.rs [hunk 2]
```diff
@@ -866,10 +866,10 @@ impl Conductor {
                                     &context,
                                     &dna_hash_from_conductor_config,
                                     &dna_hash_computed,
-                                    &dna_hash_computed_from_file, &dna_file)?;
+                                    &dna_hash_computed_from_file, &dna_location)?;
                             },
                             Err(_) => {
-                                let msg = format!("Conductor: Could not load DNA file {:?}.", &dna_file);
+                                let msg = format!("Conductor: Could not load DNA file {:?}.", &dna_location);
```
Comment (zippy) [inline crates/conductor_lib/src/conductor/base.rs]:
```suggestion
                                let msg = format!("Conductor: Could not load DNA from {:?}.", &dna_location);
```

File: crates/conductor_lib/src/config.rs [hunk 3]
```diff
@@ -236,9 +235,9 @@ impl Configuration {
                 )
             })?;
             let dna_config = dna_config.unwrap();
-            let dna =
-                Arc::get_mut(&mut dna_loader).unwrap()(&PathBuf::from(dna_config.file.clone()))
-                    .map_err(|_| format!("Could not load DNA file \"{}\"", dna_config.file))?;
+            let dna_location = dna_config.get_location();
+            let dna = Arc::get_mut(&mut dna_loader).unwrap()(&dna_location)
+                .map_err(|_| format!("Could not load DNA file \"{}\"", dna_location))?;
```
Comment (zippy) [inline crates/conductor_lib/src/config.rs]:
```suggestion
                .map_err(|_| format!("Could not load DNA from \"{}\"", dna_location))?;
```

File: crates/conductor_lib/src/config.rs [hunk 4]
```diff
@@ -345,15 +344,14 @@ impl Configuration {
...
-            Arc::get_mut(&mut dna_loader).unwrap()(&PathBuf::from(caller_dna_file.clone()))
-                .map_err(|err| {
-                    format!(
-                        "Could not load DNA file \"{}\"; error was: {}",
-                        caller_dna_file, err
-                    )
-                })?;
+            Arc::get_mut(&mut dna_loader).unwrap()(&caller_dna_file).map_err(|err| {
+                format!(
+                    "Could not load DNA file \"{:?}\"; error was: {}",
```
Comment (zippy) [inline crates/conductor_lib/src/config.rs]:
```suggestion
                    "Could not load DNA from \"{:?}\"; error was: {}",
```

File: crates/conductor_lib/src/config.rs [hunk 5]
```diff
@@ -374,15 +372,14 @@ impl Configuration {
...
-            Arc::get_mut(&mut dna_loader).unwrap()(&PathBuf::from(callee_dna_file.clone()))
-                .map_err(|err| {
-                    format!(
-                        "Could not load DNA file \"{}\"; error was: {}",
-                        callee_dna_file, err
-                    )
-                })?;
+            Arc::get_mut(&mut dna_loader).unwrap()(&callee_dna_file).map_err(|err| {
+                format!(
+                    "Could not load DNA file \"{}\"; error was: {}",
```
Comment (zippy) [inline crates/conductor_lib/src/config.rs]:
```suggestion
                    "Could not load DNA from \"{}\"; error was: {}",
```

**Comment by maackle** [comment]
@AshantiMutinta yeah Arc::get_mut is risky, but I didn't add those, they were there already and have been for a long time.

**Comment by maackle** [comment]
I'm merging this into `trycp` branch rather than develop because that's the only reason we need this right now, and I don't want to merge into develop until we've tested it a bit.

**Pull request merged**